lib/SIL/SILFunctionType.cpp

```diff
@@ -254,51 +276,72 @@ CanSILFunctionType SILFunctionType::getAutoDiffAssociatedFunctionType(
-        getAssociatedType(param.getType(), "TangentVector"),
+        getAssociatedType(param.getType(), tangentDependentType),
         param.getConvention()
       });
+    }
     SmallVector<SILResultInfo, 8> tangentResults;
     for (auto &result : getResults())
       tangentResults.push_back({
-        getAssociatedType(result.getType(), "TangentVector"),
+        getAssociatedType(result.getType(), tangentDependentType),
         result.getConvention()
       });
     auto differentialType = SILFunctionType::get(
         getGenericSignature(), ExtInfo(), SILCoroutineKind::None,
         ParameterConvention::Direct_Guaranteed, tangentParams, {},
-        tangentResults, None, getASTContext());
+        tangentResults, None, ctx);
     SmallVector<SILResultInfo, 8> jvpResults(getResults().begin(),
                                              getResults().end());
     jvpResults.push_back({differentialType, ResultConvention::Owned});
     return SILFunctionType::get(getGenericSignature(), getExtInfo(),
                                 getCoroutineKind(), getCalleeConvention(),
                                 getParameters(), getYields(), jvpResults,
-                                getOptionalErrorResult(), getASTContext());
+                                getOptionalErrorResult(), ctx);
   }
   case AutoDiffAssociatedFunctionKind::VJP: {
     SmallVector<SILParameterInfo, 8> cotangentParams;
     for (auto &result : getResults())
       cotangentParams.push_back(
           getFormalParameterInfo(
-              getAssociatedType(result.getType(), "CotangentVector")));
+              getAssociatedType(result.getType(), cotangentDependentType)));
     SmallVector<SILResultInfo, 8> cotangentResults;
-    for (auto &param : getParameters())
+    if (hasSelfParam() && testParamIndex(numParamsWithoutSelf)) {
+      auto &param = getParameters()[numParamsWithoutSelf];
       cotangentResults.push_back(
           getFormalResultInfo(
-              getAssociatedType(param.getType(), "CotangentVector")));
+              getAssociatedType(param.getType(), cotangentDependentType)));
+    }
+    for (unsigned paramIndex : range(numParamsWithoutSelf)) {
+      if (!testParamIndex(paramIndex))
+        continue;
+      auto &param = getParameters()[paramIndex];
+      cotangentResults.push_back(
+          getFormalResultInfo(
+              getAssociatedType(param.getType(), cotangentDependentType)));
+    }
     auto pullbackType = SILFunctionType::get(
         getGenericSignature(), ExtInfo(), SILCoroutineKind::None,
         ParameterConvention::Direct_Guaranteed, cotangentParams, {},
-        cotangentResults, {}, getASTContext());
+        cotangentResults, {}, ctx);
     SmallVector<SILResultInfo, 8> vjpResults(getResults().begin(),
                                              getResults().end());
     vjpResults.push_back({pullbackType, ResultConvention::Owned});
     return SILFunctionType::get(getGenericSignature(), getExtInfo(),
```

Comment:
I think this was already preserving extinfo. Your PR description seemed to indicate it wasn’t. Could you elaborate a bit more?

Reply:
Oh whoops, I meant that I made `AnyFunctionType::getAutoDiffAssociatedFunctionType` preserve extinfo. I will fix the PR description.